libraries/lib-ffmpeg-support/wrappers/AVCodecContextWrapper.cpp

```diff
@@ -0,0 +1,180 @@
+       */
+      bool flushing = packet
+         ? (packetCopy->GetSize() == 0 && packetCopy->GetData() == nullptr)
+         : true;
+      if (!flushing && packetCopy->GetData() == nullptr)
+         return {};
+
+      const int channels = GetChannels();
+
+      if (flushing)
+         GetChannels();
+
+      int bytesDecoded = 0;
+      do
+      {
+         int gotFrame;
+         // Deprecated?  https://ffmpeg.org/doxygen/3.3/group__lavc__decoding.html#gaaa1fbe477c04455cdc7a994090100db4
+         bytesDecoded = mFFmpeg->avcodec_decode_audio4(
+            mAVCodecContext, frame->GetWrappedValue(), &gotFrame,
+            packetCopy->GetWrappedValue());
+
+         if (bytesDecoded < 0)
+            return data; // Packet decoding has failed
+
+         if (gotFrame == 0)
+            /*
+             "Note that this field being set to zero does not mean that an
+             error has occurred. For decoders with AV_CODEC_CAP_DELAY set, no
+             given decode call is guaranteed to produce a frame."
+             */
+            // (Let's assume this doesn't happen when flushing)
+            continue;
+
+         const auto sampleSize =
+            static_cast<size_t>(mFFmpeg->av_get_bytes_per_sample(
+               static_cast<AVSampleFormatFwd>(frame->GetFormat())));
+
+         const auto samplesCount = frame->GetSamplesCount();
+         const auto frameSize = channels * sampleSize * samplesCount;
+
+         auto oldSize = data.size();
+         data.resize(oldSize + frameSize);
+         auto pData = &data[oldSize];
+
+         if (frame->GetData(1) != nullptr)
+         {
+            // We return interleaved buffer
+            for (int channel = 0; channel < channels; channel++)
+            {
+               for (int sample = 0; sample < samplesCount; sample++)
+               {
+                  const uint8_t* channelData =
+                     frame->GetExtendedData(channel) + sampleSize * sample;
+
+                  uint8_t* output =
+                     pData + sampleSize * (channels * sample + channel);
+
+                  std::copy(channelData, channelData + sampleSize, output);
+               }
+            }
+         }
+         else
+         {
+            uint8_t* frameData = frame->GetData(0);
+            std::copy(frameData, frameData + frameSize, pData);
+         }
+
+         packetCopy->OffsetPacket(bytesDecoded);
+      }
+      while ( flushing ? bytesDecoded > 0 : packetCopy->GetSize() > 0 );
+   }
+
+   return data;
+}
+
+namespace
+{
+unsigned int MakeTag(char a, char b, char c, char d) noexcept
+{
+   return
+      (static_cast<unsigned>(a) << 0) | (static_cast<unsigned>(b) << 8) |
+      (static_cast<unsigned>(c) << 16) | (static_cast<unsigned>(d) << 24);
+}
+}
+
+void AVCodecContextWrapper::SetCodecTagFourCC(const char* fourCC) noexcept
+{
+   if (fourCC == nullptr || std::strlen(fourCC) != 4)
+      return;
+
+   SetCodecTag(MakeTag(fourCC[0], fourCC[1], fourCC[2], fourCC[3]));
```

Comment:
I see you are correcting possibly wrong assumptions about endian-ness in the old code.

Reply:
The original code was copying random memory most of the time (as the tag was an empty string by default, so 0 and 3 random bytes were copied). So endian-ness wasn't the worst problem anyway :)